### ISSUE TITLE
communicate format output to all engines 

### ISSUE THREAD

**Issue by cscheid** [quarto-dev/quarto-cli#2216]
Via Max Kuhn:

> Does quarto have built-in environmental variables (or similar) during qmd compilation that reflects the output type? Similar to knitr::is_html_output() but cross-language and more specific about the output type?
> I’m trying to write single qmd files for books that get compiled in different formats. Any tools (e.g. chunk options etc) that make that easier would be fantastic.

**Comment by dragonstyle** [comment]
You can use conditional content at the markdown level: https://quarto.org/docs/authoring/conditional.html

The knitr functions like is_html_output should also still work.

**Comment by cscheid** [comment]
Right, but I still think we should have something that signals uniformly to Jupyter and knitr and future extensions; I think the question was posed in the context of running Python code.

**Comment by cscheid** [comment]
Reference: `notebook.py:88` is the place to do this.

**Comment by allenmanning** [comment]
Notes:  `notebook.py` ln 88

**Comment by cderv** [comment]
For reference, other mention of this in:
* https://github.com/quarto-dev/quarto-cli/discussions/8456

**Comment by cderv** [comment]
Related: 
- https://github.com/quarto-dev/quarto-cli/discussions/10173

**Comment by cderv** [comment]
For future reference, part of the discussion we had in the past was to add this information as an environment variable among the ones we set for the execution environment. So that is easy for all engine to retrieve information

https://github.com/quarto-dev/quarto-cli/blob/6173a7061d1dcea113efde2b1bc5cca56dd316d4/src/execute/environment.ts#L11-L26

**Comment by cscheid** [comment]
Yup, `ExecuteOptions` has everything we need.

**Comment by cderv** [comment]
We just need to choose the name of the environment, and which part of the format are we passing. 

I mean 
- targeted pandoc format only (`pdf`, `revealjs`, `html`, `typst`, ...) 
	- It seems necessary to produce raw content from code
- with variants if any ? 
- What about internally replaced format (like `gfm` being pandoc's `commonmark` + variants) ? 
- a separate env for Quarto own format (`dashboard` ? `email` ? `docusaurus` ?) 

Maybe we need to mimic how we split in Typescript ? (in `breakFormatString()` ? or maybe FormatIdenfier 
https://github.com/quarto-dev/quarto-cli/blob/97ffff53ef9a66bb665a128fb0cb52da4f1a6722/src/config/types.ts#L391-L396

Once we decide, it is easy to implement.

**Comment by cscheid** [comment]
I think we should send place exactly the contents of `FormatIdentifier` in an env variable.

**Comment by cderv** [comment]
Oh I see. No specific formatting, just passing the object. I forgot about JSON as message format. interesting ! I can submit something with all this information. Cool !

**Comment by cderv** [comment]
First step toward this is the base format:
- https://github.com/quarto-dev/quarto-cli/pull/13150

**Comment by cscheid** [comment]
Quarto 1.8 exports QUARTO_EXECUTE_INFO, an env var that points to a JSON file with ample execution information, including format (it's in fact the data in the `ExecuteOptions` object mentioned above)

**Comment by mcanouil** [comment]
This works great!

````
```{r}
jsonlite::read_json(Sys.getenv("QUARTO_EXECUTE_INFO"))[["format"]][["identifier"]][["base-format"]]
```
````

Also, all the other information such as the `fig-` and `languages` will be very useful.

Edit: here is what I had in mind with this (and it works!!!🤯)

````qmd
---
format: html
lang: fr
language:
  en:
    custom-title: Alpha
  fr:
    custom-title: Beta
---

```{r}
jsonlite::read_json(
  Sys.getenv("QUARTO_EXECUTE_INFO")
)[["format"]][["language"]][["custom-title"]]
```
````

**Comment by cderv** [comment]
I believe this issue was about make something like `knitr::is_html_output()` available to all engines, and closer to Quarto's format. 

Exposing all information in `QUARTO_EXECUTE_INFO` is really cool because it offers a lot of information to users and developers of tools; however, this means two things to me

- `QUARTO_EXECUTE_INFO` content should be documented properly if we expect users to use it correctly. Just on format specifiers, some fields are Quarto-specific (e.g., `base-format` vs. `target-format`). Probably part of our more advanced doc.

- Both the **quarto** R package and the **quarto** Python module should have helpers functions so users can access relevant information without knowing which fields to retrieve. This is especially useful to really provide an `is_html_output()` equivalent, or a more generic `quarto_to()`. [quarto-dev/quarto-r#285](https://github.com/quarto-dev/quarto-r/issues/285) tracks this idea.
